analysis/src/npm.js

```diff
@@ -0,0 +1,160 @@
+'use strict';
+
+const npm = require('npm');
+const childProcess = require('child_process');
+const path = require('path');
+const fs = require('fs');
+
+const Ana = require('./ana_log');
+
+const fatalErrorCodes = [];
+
+var npmBin = require.resolve('../node_modules/npm/bin/npm-cli.js');
+var nodeBin = process.env.npm_node_execpath || process.env.NODE || process.execPath;
+
+/**
+ * Service for communicating with Bower on the local machine.
+ * Provides support for installing packages, enumerating their dependencies etc...
+ */
+class NPM {
+  constructor() {
+    // Create a safe environment for us to install so dependencies of this app
+    // aren't affected by new packages.
+    if (!fs.existsSync('installed')) {
+      fs.mkdirSync('installed');
+    }
+    npm.load();
+  }
+
+  _exec(cmd, opts) {
+    return new Promise((resolve, reject) => {
+      cmd = [npmBin].concat(cmd);
+      opts = opts || {};
+      opts.env = process.env;
+
+      // Writing to file seems to prevent the stdout stream from being cut off
+      // unexpectedly.
+      const out = fs.openSync('./out.log', 'w');
+      opts.stdio = ['ignore', out, null];
+
+      var stderr = '';
+      var child = childProcess.spawn(nodeBin, cmd, opts);
+
+      if (child.stderr) {
+        child.stderr.on('data', function(chunk) {
+          stderr += chunk;
+        });
+      }
+
+      child.on('error', reject);
+
+      child.on('close', function(code) {
+        fs.close(out);
+        const stdout = fs.readFileSync('./out.log');
+        fs.unlinkSync('./out.log');
+        resolve([code, stdout, stderr]);
+      });
+    });
+  }
+
+ /**
+  * Clean up (rm -rf) the working area.
+  * @return {Promise} A promise handling the prune operation.
+  */
+  prune() {
```

Comment:
Just FYI, https://github.com/normalize/mz is a library that promisifies the node fs libs. In combination with async/await, it can be really nice.